### PR TITLE
Discretization Algorithm MDLP added

### PR DESCRIPTION
**Pull request by MarvinBleier**
Added MDLP by Fayyad and Irani. This supervised discretizer works with entropy and minimum discription length. 
1. Sort all values
2. Determine all useful cut points (usefull if the class changes at this point)
3. Search for cut point  which splits the values into the best "bi-partition". 
4. Repeat 3 for each part until no improvement is possible (recursive)

### CODE REVIEW

**Comment by MarvinBleier** [comment]
Will Refactor equalClassSplit Method to abstract Discretizer

File: DefaultConfigsExtension/src/main/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/AmevaDiscretizer.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public AmevaDiscretizer() {
         }
 
         keyValuePairs = IntStream.range(0, values.length)
-                .mapToObj(i -> new AbstractMap.SimpleImmutableEntry<>((Double) values[i], labels[i]))
+                .mapToObj(i -> new AbstractMap.SimpleImmutableEntry<>(((Number) values[i]).doubleValue(), labels[i]))
```
Comment (fkoehne) [inline DefaultConfigsExtension/src/main/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/AmevaDiscretizer.java:56]: Why the generalization / downcast?

File: DefaultConfigsExtension/src/main/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/MDLPDiscretizer.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+import java.io.Serializable;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class MDLPDiscretizer extends AbstractDiscretizer {
+
+    private List<AbstractMap.SimpleImmutableEntry<Double, Double>> keyValuePairs;
+    private List<Integer> potentialCutPoints = new ArrayList<>();
```
Comment (fkoehne) [inline DefaultConfigsExtension/src/main/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/MDLPDiscretizer.java]: Let's initialize all these with a reasonable guess regarding initial capacity (i.e. 100 or more).

File: DefaultConfigsExtension/src/test/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/MDLPDiscretizerTest.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class MDLPDiscretizerTest {
+
+    @Test
+    void testMDLPBasicDiscretization(){
```
Comment (fkoehne) [inline DefaultConfigsExtension/src/test/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/MDLPDiscretizerTest.java:16]: Let's add a test case with _only_ negative values. ;)

**Review by fkoehne — APPROVED** [review]

**Comment by fkoehne** [comment]
Also: 7.8% Duplicated Lines on New Code (is greater than 3%)... What about a util class or strategy pattern?

File: DefaultConfigsExtension/src/main/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/AmevaDiscretizer.java (hunk 1 above)
Comment (MarvinBleier) [reply DefaultConfigsExtension/src/main/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/AmevaDiscretizer.java:56]: If all values were of Integer, a castingexception would occur, because casting from Integer to Double ( values[i] gives Integer), is not possible.

Comment (MarvinBleier) [reply DefaultConfigsExtension/src/main/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/AmevaDiscretizer.java:56]: After the doublevalue of the Number-Type is taken, it will be converted to Double, i did not change the key and value Types of the AbstractMap

File: DefaultConfigsExtension/src/test/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/MDLPDiscretizerTest.java (hunk 3 above)
Comment (MarvinBleier) [reply DefaultConfigsExtension/src/test/java/de/viadee/xai/anchor/adapter/tabular/discretizer/impl/MDLPDiscretizerTest.java:16]: Added a test and the discretization works as expected.

**Comment by MarvinBleier** [comment]
> Also: 7.8% Duplicated Lines on New Code (is greater than 3%)... What about a util class or strategy pattern?

Will Refactor equalClassSplit Method to abstract Discretizer. I would like to talk to @TobiasGoerke or @fkoehne about it first.

**Review by TobiasGoerke — APPROVED** [review]
